{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+1 57166..61908  4743 <br>\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra Credit - 10 points possible\n",
    " - Allow your program to optionally use a set of start codons - ATG, GTG, TTG for example.\n",
    " - Allow your program to optionally use a set of stop codons - TAG, TGA, TAA for example.\n",
    " - Allow your program to print every putative gene in an ORF instead of only the largest.\n",
    " - Allow your program to have varying ORF size minimums \n",
    "\n",
    "# Submit your code and answers\n",
    "For this lab, you should upload the following three files as attachments:\n",
    "\n",
    " - your sequenceAnalysis.py module\n",
    " - findORFs.py, include design in # comments at the beginning of findORFs.py\n",
    " - tass2ORFdata-ATG-100.txt\n",
    "\n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Include an overview about what your program is designed to do with expected inputs and outputs\n",
    " - Include design (place at the top of your file below your name, group, and description). Include any assumptions or design decisions you made in writing your code\n",
    " - Adhere to the Report format specification\n",
    " - Include docstrings for Program, Module, Class, Methods and Functions\n",
    " - Contain in-line comments\n",
    " \n",
    " Congratulations, you finished your fifth lab assignment!\n",
    "\n",
    "# Hints:\n",
    "## One strand at a time\n",
    "\n",
    "One design solution considers only the top strand to generate gene candidates. We then generate the reverse complement and a second list of gene candidates.  When done in this way, remember to fix the start/end coordinates that you get from the reverse complement solution, since those will be based on the other end of the sequence. \n",
    "\n",
    "An alternative design exists for the reverse strand that does not require generation of the reverse complement. This solution is interesting and may be simpler to consider. This solution scans the bottom strand left to right, finding stop codons first (reverse complement), then looking for start codons (reverse complement).\n",
    "\n",
    "A third solution for the bottom strand would scan right to left, finding starts until a stop is encountered.\n",
    "\n",
    "## Start and Stop handling\n",
    "\n",
    "If you are going to do the extra-credit, then you will need to remember where the starts are until you find a stop.  This is straightforward if you place those start positions in a list, organized by frame. Even if you are not doing the extra-credit, placing starts on a list organized by frame will really simplify your code. This solution would require three lists for each of the three reading frames.\n",
    "\n",
    "As an alternative, you could scan each of the three reading frames seperately, counting by threes in each case. This seems to be the most straightforward way to do this assignment.\n",
    "\n",
    "When you do find a stop, you then consider the start(s) that you found along the way in this ORF.  Notice that the \"longest\" ORF is then defined by the start that is at the beginning of your start list.  Don't forget to clean up this list after you find genes using this stop since it would terminate all of those \"pending\" genes. What would happen if you didn't clean up your start list - you would be finding ORF candidates that had a stop in the middle of them, which means they are not __OPEN__ reading frames.\n",
    "\n",
    "## Termini and gene fragments\n",
    "\n",
    "This is a bit easier if you think that position 0(1 externally) will allways be a start position for a gene in every frame. This is the case where a start codon exists somewhere upstream of our sequence. It would code for a gene that would terminate in our sequence somewhere. This is the classic example of a dangling-stop. We would report this gene fragment beginning at position 0(1) and ending at the end of that in-frame stop. In frame 1, there might actually be a start codon at this position so dont count it twice. In Frames 2 or 3, there are extra bases ahead of our first full codon, so we will always have the possibility that the actual start is upstream of this dna fragment, so.. we report this as a dangling stop case starting at position 0(1).  \n",
    "\n",
    "Dangling starts always end at the end of sequence, even if that position is out of frame. In cases where no start or stop codon is found anywhere in the frame, the entire sequence is then considered as a gene.\n",
    "\n",
    "# Data and output files\n",
    "\n",
    "The cell below is set up for the non-extra credit version of the program with a fixed filename. Your much more usable program will use STDIN and STDOUT in command line fashion. To do this, simplify main so your fastaReader object is reading from STDIN.\n",
    "\n",
    "All of the input files and programs should be in the same folder so you wonâ€™t have to specify a path, if your data file is in some other folder, you can provide that like this:<br> ~/Desktop/bme160/someOtherPlace/labData.fa.\n",
    "\n",
    "# Comments\n",
    "\n",
    "To get full credit for this lab, you must turn in your design for each program you submit. Remember that any item you write or include must have proper docstrings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Brandon Wong (bwong44)\n",
    "# Group Members: Tim Lee, Vibha Rao, Penetha Jayakumar\n",
    "\n",
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "class CommandLine() :\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    " \n",
    "    '''\n",
    "    \n",
    "    def __init__(self, inOpts=None) :\n",
    "        '''\n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "        \n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does', \n",
    "                                             epilog = 'Program epilog - some other stuff you feel compelled to say', \n",
    "                                             add_help = True, #default is True \n",
    "                                             prefix_chars = '-', \n",
    "                                             usage = '%(prog)s [options] -option1[default] <input >output'\n",
    "                                             )\n",
    "        self.parser.add_argument('-lG', '--longestGene', action = 'store', nargs='?', const=True, default=False, help='longest Gene in an ORF')\n",
    "        self.parser.add_argument('-mG', '--minGene', type=int, choices= (100,200,300,500,1000), default=100, action = 'store', help='minimum Gene length')\n",
    "        self.parser.add_argument('-s', '--start', action = 'append', default = ['ATG'],nargs='?', \n",
    "                                 help='start Codon') #allows multiple list options\n",
    "        self.parser.add_argument('-t', '--stop', action = 'append', default = ['TAG','TGA','TAA'],nargs='?', help='stop Codon') #allows multiple list options\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')  \n",
    "        if inOpts is None :\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "########################################################################\n",
    "# OrfFinder\n",
    "# Here is the OrfFinder program\n",
    "# \n",
    "#\n",
    "########################################################################\n",
    "\n",
    "class OrfFiner:\n",
    "    \"\"\"\n",
    "    Defines a class that reads the sequence and finds the open reading frames.\n",
    "\n",
    "    Iterates 3 times for each frame\n",
    "        Iterates for each codon stepping by 3 characters\n",
    "            If has start codon:\n",
    "                Append to start codon list\n",
    "            If has stop codon:\n",
    "                For items in start codon list:\n",
    "                    Appends to geneFragList (frame,start,stop,length)\n",
    "                Clears start codon list\n",
    "            If there has been no start codons but found a stop codon we have found a dangling stop:\n",
    "                Appends to geneFragList(frame,start,stop,length)\n",
    "                Clears start codon list\n",
    "        If has start but no stop we have found a dangling start:\n",
    "            Appends to geneFragList(frame,start,stop,length)\n",
    "        If doesn't have start, stop we have found a dangling start stop:\n",
    "            Append to geneFragList(frame,start,stop,length)\n",
    "        adjustFrame += 1\n",
    "    \n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        \"\"\"Initializes the OrfFinder class and creates dictionaries that stores the ORF data.\"\"\"\n",
    "        self.startSet = set()\n",
    "        self.stopSet = set()\n",
    "        self.orfList = [] #Holds start positions\n",
    "        self.geneFragList = [] #[reading frame, start, end, length]\n",
    "        self.minOrf = 100\n",
    "        self.longestGene = False\n",
    "\n",
    "    def readFrame(self, seq):\n",
    "        \"\"\"Accepts the sequence as a string from fasta and parses it for the ORFs, looking for gene fragments.\"\"\"\n",
    "        adjustFrame = 0 #Stores value of which reading frame it is in 'TAG','TGA','TAA\n",
    "        danglingStartStop = []\n",
    "        hasStart = 0\n",
    "        hasCodon = 0\n",
    "        for orf in range(3): #Iterates for the 3 reading frames\n",
    "            frameOne = range(adjustFrame,len(seq),3)\n",
    "            hasStart = 0\n",
    "            hasStop  = 0\n",
    "            self.orfList.clear()\n",
    "            for frame in frameOne: #Iterates through the sequence 3 codons at a time\n",
    "                if seq[frame:frame+3] in self.startSet: #Checks for start\n",
    "                    self.orfList.append(frame + 1)\n",
    "                    hasStart = 1\n",
    "\n",
    "                elif seq[frame:frame+3] in self.stopSet and hasStart: #Checkts for stop\n",
    "                    for pos in self.orfList:\n",
    "                        self.geneFragList.append([adjustFrame + 1,pos,frame+3,frame+3 - pos + 1])\n",
    "                    hasStop = 1\n",
    "                    self.orfList.clear()\n",
    "\n",
    "                elif hasStart == 0 and seq[frame:frame+3] in self.stopSet: #Checks for dangling stop (we find stop but no start)\n",
    "                    self.geneFragList.append([adjustFrame+1,1,frame+3,frame+3])\n",
    "                    hasStop = 1\n",
    "                    self.orfList.clear()\n",
    "\n",
    "            if hasStart == 1 and hasStop == 0: #Checks for dangling start (we find start but no stop)\n",
    "                self.geneFragList.append([adjustFrame+1,1,len(seq),len(seq)])\n",
    "            elif hasStart == 0 and hasStop == 0 and self.orfList == []: #Checks for dangling start and stop \n",
    "                self.geneFragList.append([adjustFrame+1,1,len(seq),len(seq)])\n",
    "            adjustFrame +=1                 \n",
    "\n",
    "    def readReverseComp(self, seq):\n",
    "        \"\"\"Takes the reverse complement of the sequuence and searches for valid gene fragments\"\"\"\n",
    "        complementDict = {\"A\":\"T\",\"T\":\"A\",\"G\":\"C\",\"C\":\"G\",}\n",
    "        seqReverse = seq[::-1]\n",
    "        reverseComplement = \"\".join(complementDict[nucleotide] for nucleotide in seqReverse)\n",
    "        adjustFrame = 0 #Stores value of which reading frame it is in 'TAG','TGA','TAA\n",
    "        negativeFrame = 0 #Stores negative frame value\n",
    "        hasStart = 0\n",
    "        hasCodon = 0\n",
    "        for orf in range(3): #Iterates for the 3 reading frames\n",
    "            frameOne = range(adjustFrame,len(reverseComplement),3)\n",
    "            hasStart = 0\n",
    "            hasStop  = 0\n",
    "            self.orfList.clear()\n",
    "            for frame in frameOne: #Iterates through the sequence 3 codons at a time\n",
    "                if reverseComplement[frame:frame+3] in self.startSet: #Checks for start\n",
    "                    self.orfList.append(frame + 1)\n",
    "                    hasStart = 1\n",
    "\n",
    "                elif reverseComplement[frame:frame+3] in self.stopSet and hasStart: #Checkts for stop\n",
    "                    for pos in self.orfList:\n",
    "                        self.geneFragList.append([negativeFrame-1,len(reverseComplement)-pos+1,len(reverseComplement)-frame-2,frame+3 - pos + 1])\n",
    "                    hasStop = 1\n",
    "                    self.orfList.clear()\n",
    "\n",
    "                elif hasStart == 0 and reverseComplement[frame:frame+3] in self.stopSet: #Checks for dangling stop (we find stop but no start)\n",
    "                    self.geneFragList.append([negativeFrame-1,len(reverseComplement),len(reverseComplement)-frame-2,frame+3])\n",
    "                    hasStop = 1\n",
    "                    self.orfList.clear()\n",
    "\n",
    "            if hasStart == 1 and hasStop == 0: #Checks for dangling start (we find start but no stop)\n",
    "                self.geneFragList.append([negativeFrame-1,1,len(reverseComplement),len(reverseComplement)])\n",
    "            elif hasStart == 0 and hasStop == 0 and self.orfList == []: #Checks for dangling start and stop \n",
    "                self.geneFragList.append([negativeFrame-1,1,len(reverseComplement),len(reverseComplement)])\n",
    "            adjustFrame +=1\n",
    "            negativeFrame -= 1\n",
    "        \n",
    "    def minimumOrf(self):\n",
    "        \"Excludes the genes with the length < minOrf\"\n",
    "        minLenList = []\n",
    "        for gene in range(len(self.geneFragList)): #Iterates through the list of gene fragments\n",
    "            if self.geneFragList[gene][3] > self.minOrf: #Checks if the gene fragment is greater than the min\n",
    "                minLenList.append(self.geneFragList[gene])\n",
    "        self.geneFragList = minLenList\n",
    "\n",
    "    def sortBySize(self):\n",
    "        \"\"\"Takes the list of gene fragments and sorts by the largest gene size then starting number.\"\"\"\n",
    "        sortedGeneList = sorted(self.geneFragList, key = lambda x: (-x[-1],x[1])) #Sorts by gene length\n",
    "        self.geneFragList = sortedGeneList\n",
    "\n",
    "    def largestOrf(self):\n",
    "        \"Returns only the largest genes if the option is set.\"\n",
    "        if self.longestGene == True: #Checks if longest gene is written in terminal\n",
    "            largestList = []\n",
    "            compareSet = set()\n",
    "            for gene in self.geneFragList: #Compares the ending position since it is already sorted by largest gene\n",
    "                if gene[2] not in compareSet:\n",
    "                    largestList.append(gene)\n",
    "                    compareSet.add(gene[2])\n",
    "            self.geneFragList = largestList\n",
    "        \n",
    "    def reverseOrder(self):\n",
    "        \"\"\"Reverses the order so that the start and stop match the top strand.\"\"\"\n",
    "        for gene in self.geneFragList:\n",
    "            if gene[0] < 0:\n",
    "                startHolder = gene[1]\n",
    "                stopHolder = gene[2]\n",
    "                gene[1] = stopHolder\n",
    "                gene[2] = startHolder\n",
    "            \n",
    "    def clearGeneFragList(self):\n",
    "        \"\"\"Clears out the geneFragList object so that it doesn't store the data of previous sequences if there are any.\"\"\"\n",
    "        self.geneFragList = []\n",
    "\n",
    "########################################################################\n",
    "# Main\n",
    "# Here is the main program\n",
    "# \n",
    "#\n",
    "########################################################################\n",
    "   \n",
    "def main(inFile = None, options = None):\n",
    "    \"\"\"Finds all valid gene fragments using the terminal to specify options.\"\"\"\n",
    "    thisCommandLine = CommandLine(options)\n",
    "    reader = FastaReader(inFile)\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python3 findOrfs.py  <somefile  >someOtherFile -mG 300 -lG -s ATG -s GTG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs output using tass2.fa  -mG=300 -lG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tass2 NODE_159_length_75728_cov_97.549133\n",
    "+1 57166..61908  4743\n",
    "-1  8192..11422  3231\n",
    "+2 65963..69004  3042\n",
    "-3 14589..16862  2274\n",
    "-2  2968.. 4872  1905\n",
    "+1 64093..65952  1860\n",
    "-3    30.. 1694  1665\n",
    "+1 69475..71052  1578\n",
    "+1 48805..50223  1419\n",
    "+1 47398..48798  1401\n",
    "-3 29133..30500  1368\n",
    "-1 40922..42250  1329\n",
    "-2 19975..21270  1296\n",
    "+3 72273..73559  1287\n",
    "-1 24482..25639  1158\n",
    "-1  6689.. 7804  1116\n",
    "+1 50251..51366  1116\n",
    "-3 27501..28601  1101\n",
    "+2 63038..64078  1041\n",
    "+3 62019..63038  1020\n",
    "-2 42271..43263   993\n",
    "+3 51864..52805   942\n",
    "+1 45484..46371   888\n",
    "-3 11433..12317   885\n",
    "+2 74357..75184   828\n",
    "+2 71576..72394   819\n",
    "+3 46341..47138   798\n",
    "-2 18613..19407   795\n",
    "+1 55642..56388   747\n",
    "-1 16940..17632   693\n",
    "-2 13288..13974   687\n",
    "-3 26115..26801   687\n",
    "-1 21338..21994   657\n",
    "-1 30998..31654   657\n",
    "-2 12601..13251   651\n",
    "-2  4894.. 5532   639\n",
    "-3 32592..33221   630\n",
    "-1 39914..40525   612\n",
    "+1 53977..54588   612\n",
    "+3 54588..55193   606\n",
    "-3 33234..33809   576\n",
    "-3 22002..22559   558\n",
    "-3 23859..24413   555\n",
    "-2  1945.. 2490   546\n",
    "+1 73861..74370   510\n",
    "-2  6214.. 6696   483\n",
    "+1 16324..16806   483\n",
    "-1 22556..23032   477\n",
    "+3 32808..33260   453\n",
    "+2 53324..53776   453\n",
    "+1 29056..29502   447\n",
    "-2 36286..36729   444\n",
    "+3 51396..51833   438\n",
    "+2 55196..55627   432\n",
    "-1  2468.. 2896   429\n",
    "-2 31798..32220   423\n",
    "+1 52891..53307   417\n",
    "-2 30595..30996   402\n",
    "-2  5809.. 6204   396\n",
    "+1     1..  393   393\n",
    "-2 35740..36129   390\n",
    "-3 34542..34925   384\n",
    "+2 56474..56857   384\n",
    "-1 17888..18268   381\n",
    "-3 23004..23381   378\n",
    "-3 52968..53336   369\n",
    "-1 32207..32572   366\n",
    "-1 23396..23752   357\n",
    "+2 36419..36775   357\n",
    "-2 18268..18609   342\n",
    "+2 44594..44935   342\n",
    "+1 43714..44049   336\n",
    "+2  6557.. 6886   330\n",
    "-2 34927..35253   327\n",
    "-3 34131..34451   321\n",
    "+1 30778..31095   318\n",
    "+2 24974..25288   315\n",
    "+1  1246.. 1557   312\n",
    "-3 35253..35561   309\n",
    "-2 39160..39468   309\n",
    "-2 68932..69234   303"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs output using lab5test.fa at mG=0 cutoff\n",
    "\n",
    "(Note: I have included the sequence of the gene candidate as an additional column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test\n",
    "+1     1..    9     9 ATGATGTAA\n",
    "+3     1..    9     9 AT GAT GTA A\n",
    "-1     1..    9     9 TTACATCAT\n",
    "-2     1..    9     9 T TAC ATC AT\n",
    "-3     1..    9     9 TTACATCAT\n",
    "+1     4..    9     6 ATG TAA\n",
    "+2     1..    4     4 ATGA\n",
    "test2\n",
    "+1     1..   10    10 AA TGA TGT AA\n",
    "+2     1..   10    10 A ATG ATG TAA\n",
    "-1     1..   10    10 TTACATCATT\n",
    "-2     1..   10    10 TTACATCATT\n",
    "-3     1..   10    10 TTACATCATT\n",
    "+2     2..   10     9 ATGATGTAA\n",
    "+2     5..   10     6 ATGTAA\n",
    "+3     1..    5     5 AATGA\n",
    "test3\n",
    "+2     1..   11    11 A AAT GAT GTA A\n",
    "+3     1..   11    11 AA ATG ATG TAA\n",
    "-1     1..   11    11 TTACATCATTT\n",
    "-2     1..   11    11 TTACATCATTT\n",
    "-3     1..   11    11 TTACATCATTT\n",
    "+3     3..   11     9 ATG ATG TAA\n",
    "+1     1..    6     6 AAATGA\n",
    "+3     6..   11     6 ATGTAA\n",
    "test-1\n",
    "+1     1..    9     9 TTACATCAT\n",
    "+2     1..    9     9 TTACATCAT\n",
    "+3     1..    9     9 TTACATCAT\n",
    "-1     1..    9     9 ATGATGTAA\n",
    "-3     1..    9     9 ATGATGTAA\n",
    "-1     1..    6     6 ATGTAA\n",
    "-2     6..    9     4 ATGA\n",
    "test-2\n",
    "+1     1..   10    10 TTACATCATA\n",
    "+2     1..   10    10 TTACATCATA\n",
    "+3     1..   10    10 TTACATCATA\n",
    "-1     1..   10    10 TATGATGTAA\n",
    "-2     1..   10    10 TATGATGTAA\n",
    "-2     1..    9     9 ATGATGTAA\n",
    "-2     1..    6     6 ATGTAA\n",
    "-3     6..   10     5 TATGA\n",
    "test-3\n",
    "+1     1..   11    11 TTACATCATAA\n",
    "+2     1..   11    11 TTACATCATAA\n",
    "+3     1..   11    11 TTACATCATAA\n",
    "-2     1..   11    11 TTATGATGTAA\n",
    "-3     1..   11    11 TTATGATGTAA\n",
    "-3     1..    9     9 ATGATGTAA\n",
    "-3     1..    6     6 ATGTAA\n",
    "-1     6..   11     6 TTATGA\n",
    "test1A\n",
    "+3     1..   10    10 ATGATGTAAA\n",
    "-1     1..   10    10 TTTACATCAT\n",
    "-2     1..   10    10 TTTACATCAT\n",
    "-3     1..   10    10 TTTACATCAT\n",
    "+1     1..    9     9 ATGATGTAA\n",
    "+1     4..    9     6 ATGTAA\n",
    "+2     1..    4     4 ATGA\n",
    "test2A\n",
    "+1     1..   11    11 AATGATGTAAA\n",
    "-1     1..   11    11 TTTACATCATT\n",
    "-2     1..   11    11 TTTACATCATT\n",
    "-3     1..   11    11 TTTACATCATT\n",
    "+2     1..   10    10 AATGATGTAA\n",
    "+2     2..   10     9 ATGATGTAA\n",
    "+2     5..   10     6 ATGTAA\n",
    "+3     1..    5     5 AATGA\n",
    "test3A\n",
    "+2     1..   12    12 AAATGATGTAAA\n",
    "-1     1..   12    12 TTTACATCATTT\n",
    "-2     1..   12    12 TTTACATCATTT\n",
    "-3     1..   12    12 TTTACATCATTT\n",
    "+3     1..   11    11 AAATGATGTAA\n",
    "+3     3..   11     9 ATGATGTAA\n",
    "+1     1..    6     6 AAATGA\n",
    "+3     6..   11     6 ATGTAA\n",
    "test-1A\n",
    "+1     1..   10    10 ATTACATCAT\n",
    "+2     1..   10    10 ATTACATCAT\n",
    "+3     1..   10    10 ATTACATCAT\n",
    "-3     1..   10    10 ATGATGTAAT\n",
    "-1     2..   10     9 ATGATGTAA\n",
    "-1     2..    7     6 ATGTAA\n",
    "-2     7..   10     4 ATGA\n",
    "test-2A\n",
    "+1     1..   12    12 AATTACATCATA\n",
    "+2     1..   12    12 AATTACATCATA\n",
    "+3     1..   12    12 AATTACATCATA\n",
    "-1     1..   12    12 TATGATGTAATT\n",
    "-2     3..   12    10 TATGATGTAA\n",
    "-2     3..   11     9 ATGATGTAA\n",
    "-2     3..    8     6 ATGTAA\n",
    "-3     8..   12     5 TATGA\n",
    "test-3A\n",
    "+1     1..   13    13 AATTACATCATAA\n",
    "+2     1..   13    13 AATTACATCATAA\n",
    "+3     1..   13    13 AATTACATCATAA\n",
    "-2     1..   13    13 TTATGATGTAATT\n",
    "-3     3..   13    11 TTATGATGTAA\n",
    "-3     3..   11     9 ATGATGTAA\n",
    "-3     3..    8     6 ATGTAA\n",
    "-1     8..   13     6 TTATGA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Inspection Results:\n",
    "\n",
    "#Vibha inspection:\n",
    "I have no comments. The code looks great! I like how you've implemented the longestOrf method. \n",
    "What does the reverseOrder method do?\n",
    "\n",
    "#Tim Inpsection:\n",
    "You have some extra commented code that you can clean up.\n",
    "\n",
    "#Response:\n",
    "For my reverseOrder method it iterates through all the gene fragments and searches for the negative frames. Once it finds a negative frame it changes the start and stop position to the correct \n",
    "position to be in the same orientation as the top strand.\n",
    "\n",
    "I took out the exta lines to clean up my code."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
