{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find Unique\n",
    " - Deliverables:\n",
    "     - findUnique.py - 50 points ( We will test this code)\n",
    "     - Notebook with Inspection materials and Results ( 5 pts )\n",
    "     - required input from STDIN\n",
    "     - required output to STDOUT\n",
    "     - output must be sorted by tRNA header\n",
    "\n",
    "## Sets and tRNA\n",
    "\n",
    "Mitochondrial tRNA are encoded in the mitochondrial genome and account for ~10% of the coding space, yet over 50% of mitochondrial genomic disease have their origin in this molecule. These molecules are transcribed, processed, modified, and ultimately folded to produce these mature adapters between the mitochondrial transcription and translation processes. Defects in mt.tRNA maturation reflect additional disease states .. if only we could identify those mutations. We are working on such a device, though we first need to use it to identify abundance of these molecules in a mixed population. Are there any unique subsequences among the 22 mt.tRNA that can be used as \"tags\"? If so, we can count those tags to assess abundance.\n",
    "\n",
    "## Assignment\n",
    "\n",
    "Write a python command line program that reads a file of fasta sequences from STDIN, finds the unique subsequences that occur in each single tRNA such that no members of this set occur among any of the other tRNA sets. Each of those 22 sets should be minimized, such that no member of that unique subsequence set is a substring of any other member of this set. [ Unique and Essential]\n",
    "\n",
    "As an example, let's say that both ACG and AAACGA are in a unique set. Since ACG is a substring of AAACGA we would remove AAACGA. [ ACG is Essential ]\n",
    "\n",
    "Use Python sets for this assignment[__required__]. Not only will your code be smaller, but it will be more likely to work. The union, intersection and difference operators will be quite useful.\n",
    "\n",
    "## Rough design plan...\n",
    "0) Consider a tRNA as an object. In thi scase we would then have 22 objects that woul dneed to be able to communicate with each other.\n",
    "\n",
    "1) compute the set of all substrings from each tRNA sequence. [powerset] I will refer to a set of substrings as a tRNAset.\n",
    "\n",
    "2) for each tRNAset, compute the union of all other tRNA sets, and then remove that union from the current tRNAset. Notice that this union operation finds all of the substrings from all other tRNA. If any of those are present in your current tRNA, then they are not unique ! [ Unique]\n",
    "\n",
    "3) for each unique tRNAset, it now contains the truly unique ones, along with any extensions of that subsequence. If, for example, it was found that G only occurred in a single tRNA, then adding an A onto that G must also be unique because it has a G in it. We only want the minimal form.. G. [ Essential]\n",
    "\n",
    "4) Remove spaces from the header line before sorting and printing. This will make your output a little prettier.\n",
    "\n",
    "5) make sure to remove any alignment characters from your initial sequences. These characters are periods(.), underscores(\\_), or dashes(-) . You will find many new characters in the sequence other than {ACGU} - leave these in place, they are modified bases.\n",
    "\n",
    "## Report\n",
    "Print a report that contains items as follows. \n",
    "\n",
    " - Line 1: the tRNA name\n",
    " - Line 2: the tRNA sequence ( with the alignment characters removed)\n",
    " - lines 3-80 or so, each unique element.\n",
    "\n",
    "These unique elements need to be ordered by their starting position in the tRNA sequence, and should be spaced over so they are directly under that subsequence in the original tRNA. This looks like an alignment, but you can find where it belongs by using the string.find() method. Include dots in all positions to the left, serving as alignment guides for your reader. [ see sample output below ]\n",
    "\n",
    "Do this for all of the 22 tRNA sequences.\n",
    "\n",
    "Print the tRNA out as above, sorted by the header line.  \n",
    "\n",
    "## Hints:\n",
    "\n",
    "__use sets and the set operators - this is required !__\n",
    "\n",
    "your final code will be under 100 lines.\n",
    "\n",
    "Do most of your coding using methods defined in a class and then instantiated as objects. \n",
    "\n",
    "The sequences include characters that are just alignment characters. They are not part of the sequence and must be removed. [-\\_\\.] are alignment characters. \n",
    "\n",
    "When removing items from a tRNAset, don't do this while iterating through that set. Also, when building a unique set you will need the original contents of all other tRNAsets. So.. build a new set to keep the unique contents, or keep track of the elements you intend to delete later. Notice that you build the union of all other tRNA, and this happens 22 times - these unions are all distinct from each other. Example, consider 4 sets, A,B,C,D.  we would compute the set B, C, D to use against set A, and we would compute the set A, C, D to use against B. A and B need to not change while we are doing this computation.\n",
    "\n",
    "There are cases where a unique element is present multiple times in a single tRNA. This sequence is unique to this tRNA and should be reported in its multiple locations ( same sequence on multiple lines of the output).\n",
    "\n",
    "This is a command line program, though it does not have any optional parameters. You really don't need the commandline class. Your input comes from STDIN and output goes to STDOUT. Use the FastaReader for input and print statement for output. If you are going to use jupyter to develop your code, you can add a filename to fastareader for your testing.\n",
    "\n",
    "## Submission\n",
    "\n",
    "Submit code using Canvas. As always, work with your group and write your own code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Brandon Wong (bwong44)\n",
    "# Group Members: Tim Lee, Vibha Rao, Penetha Jayakumar\n",
    "\n",
    "import sequenceAnalysis as sa\n",
    "\n",
    "########################################################################\n",
    "# findUnique Class\n",
    "# \n",
    "# \n",
    "########################################################################\n",
    "\n",
    "class findUnique:\n",
    "    \"\"\"Finds unique subsequences across the different tRNAs.\"\"\"\n",
    "    InstanceList = [] #Holds all the tRNA objects\n",
    "    \n",
    "    def __init__(self):\n",
    "        \"\"\"Initializes the findUnique class.\"\"\"\n",
    "        self.powerset = set() #Holds powerset of the sequence\n",
    "        self.head = \"\" #Holds the header line of the sequence\n",
    "        self.essential = set()  #Holds essential set of sequence\n",
    "        self.sequence = \"\" #Initial sequence without alignment characters\n",
    "        findUnique.InstanceList.append(self) #Appends the tRNA object\n",
    "\n",
    "    def cleanSeq(self, seq):\n",
    "        \"\"\"Cleans the sequence of any alignment characters from initial sequences.\"\"\"\n",
    "        cleanedSeq = seq.replace(\".\",\"\").replace(\"_\",\"\").replace(\"-\",\"\")\n",
    "        self.sequence = cleanedSeq\n",
    "        return cleanedSeq\n",
    "    \n",
    "    def createPowerset(self, seq):\n",
    "        \"\"\"Creates powerset for each sequence.\"\"\"\n",
    "        for i in range(len(seq)): #Nested loop to get all subsequences\n",
    "            for j in range(i+1, len(seq)+1): #Accouts for slicing exclusion\n",
    "                self.powerset.add(seq[i:j])\n",
    "\n",
    "    def findUnique(self):\n",
    "        \"\"\"Finds the unique subsequences in the 22 tRNAs.\"\"\"\n",
    "        unionSet = set()\n",
    "        copyInstanceList = findUnique.InstanceList.copy()\n",
    "        copyInstanceList.remove(self)\n",
    "        for tRNA in copyInstanceList: #Creates union of all other sets\n",
    "            unionSet = unionSet.union(tRNA.powerset)\n",
    "        uniqueSet = self.powerset.difference(unionSet) #Differnce of self vs union of all the others\n",
    "        #Find Essentials\n",
    "        self.essential = uniqueSet.copy() #Copy set so we don't run into iteration error\n",
    "        for unique in uniqueSet:\n",
    "            for i in uniqueSet:\n",
    "                try:\n",
    "                    if unique != i and unique in i:\n",
    "                        self.essential.remove(i)\n",
    "                except KeyError: #Handles errors where unique is bigger than i\n",
    "                    continue    \n",
    "\n",
    "    def alignSeq(self):\n",
    "        \"\"\"Aligns the subsequences based on where they are in the sequence.\"\"\"\n",
    "        alignList = [] #Holds a list of list, [position,essential]\n",
    "\n",
    "        for essentialSeq in self.essential: #Finds the position of where each essential segment starts\n",
    "            alignList.append([self.sequence.find(essentialSeq),essentialSeq])\n",
    "        alignList.sort()\n",
    "\n",
    "        for item in alignList:\n",
    "            print(\".\"*item[0] +item[1])\n",
    "        \n",
    "\n",
    "########################################################################\n",
    "# Main\n",
    "# Here is the main program\n",
    "# \n",
    "########################################################################\n",
    "\n",
    "def main(inCL=None):\n",
    "    \"\"\"Runs the findUnique class to find the unique subsequences of each tRNA.\"\"\"\n",
    "    reader = sa.FastAreader()\n",
    "    for head, seq in reader.readFasta(): #Iterates through the fasta file\n",
    "            tRNA = findUnique()\n",
    "            tRNA.head = head.strip()\n",
    "            tRNA.createPowerset(tRNA.cleanSeq(seq))\n",
    "\n",
    "    sortedHead = sorted(findUnique.InstanceList, key=lambda x: x.head) #Sorts by header\n",
    "    for object in sortedHead: #Iterates tRNA objects to find uniques\n",
    "        object.findUnique()\n",
    "        print(object.head)\n",
    "        print(object.sequence)\n",
    "        object.alignSeq()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tRNA|Lys|∃UU|Bostaurus|mitochondrial\n",
    "CACUAAGA\"LCUAUAUAGCACPAACCU∃UU6AGUUAGAGAUUGAGAGCCAU\"UACUCUCCUUGGUGACCA\n",
    "CACU\n",
    ".ACUA\n",
    "...UAA\n",
    "....AAG\n",
    ".......A\"\n",
    ".........L\n",
    "..........CUAU\n",
    "............AUAU\n",
    "..............AUAG\n",
    "...............UAGC\n",
    ".................GCA\n",
    ".....................P\n",
    "......................AAC\n",
    ".......................ACCU\n",
    "...........................∃\n",
    "..............................6\n",
    "...............................AGU\n",
    "................................GUU\n",
    ".................................UUA\n",
    "..................................UAGA\n",
    "...................................AGAGA\n",
    "......................................GAU\n",
    ".......................................AUU\n",
    "........................................UUGA\n",
    ".........................................UGAG\n",
    "..........................................GAGAG\n",
    "............................................GAGC\n",
    "..............................................GCC\n",
    "................................................CAU\n",
    "..................................................U\"\n",
    "...................................................\"U\n",
    "....................................................UAC\n",
    ".....................................................ACUC\n",
    ".......................................................UCU\n",
    ".........................................................UCC\n",
    "...........................................................CUU\n",
    "..............................................................GG\n",
    "...............................................................GUG\n",
    ".................................................................GAC\n",
    "..................................................................ACCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Inspection:\n",
    "\n",
    "Tim Inspection:\n",
    "Have you though about using list comp instead of a nested loop?\n",
    "\n",
    "\n",
    "Vibha comments:\n",
    "    - Your code looks great! Creating a separate method for finding essentials could have been another way of designing the code. \n",
    "\n",
    "\n",
    "Answer:\n",
    "I didn't think of using a list comprehension, but I think it could work since I'm adding to a list anways and I could convert it to a set aftefwards.\n",
    "\n",
    "I thought about creating a separate method for essentials but I was already writing the code for essentials so I kept it with finding uniques since it was already there."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
